### PR TITLE
Add Docker-based E2E test infrastructure

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

This is a first pass at adding a Docker-based E2E test infrastructure for the Site Kit plugin. This way actual user behavior can be emulated.

Addresses issue #157.

As outlined in #157, this is based on the E2E that is successfully being used in WordPress core, Gutenberg, and also the AMP plugin.

The environment can be started using `npm run env:start`, tests can be run using `npm run test:e2e`.

Tests are written using Jest in `tests/e2e/specs`.

This PR also adds one example test to verify that the `Congratulations, the Site Kit plugin is now activated` message is being shown upon plugin activation and that it leads the user to the setup screen.

`@wordpress/e2e-test-utils` is being used to have some helper functions for things like  activating plugins during tests and navigating to admin pages.

If some more specific helper functionality is used during tests, e.g. to mock an already configured site, WordPress plugins can be added to `tests/e2e/plugins` to do so. These can then be activated as usual during tests.

**Note**: While this adds lots of setup files to `bin/local-env`, this is a one-time setup thing. Plus, these  files might not be needed anymore in the future when this setup is further improved in WordPress core to be more easily reusable.

**To do**:

* [ ] Add some documentation to the wiki.

## Relevant technical choices

There were a few things not directly related to the test setup that I had to adjust:

* Bumped `core-js` version from 2 to 3, as that's what is required by the `@wordpress/scripts` utility (and all WP packages, really).  
  I did so by bumping the version and using the recommended `@wordpress/babel-preset-default` package that incorporates `@babel/preset-env` and deals with `core-js`.
* Modify Travis config to not unnecessarily install PHP test suite when running E2E tests.

## Checklist

- [x] My code is tested and passes existing tests.
- [x] My code has an appropriate set of tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -64,29 +70,55 @@ script:
...
+      name: E2E Tests (WordPress 4.7)
+      env: E2E=1 WP_VERSION=4.7.13
+      before_script:
+        - composer install
+        - npm install || exit 1
+      script:
+        - npm run build || exit 1 # Build for tests.
+        - npm run env:start
+        - npm run env:reset-site
+        - npm run test:e2e
```
Comment (felixarntz) [inline .travis.yml]: Also, shouldn't this also run `npm run env:stop` for completeness?

**Comment by swissspidy** [comment]
> Could you add a third Travis-CI job for WordPress 4.9 with the Gutenberg plugin (version 4.9) active? Or would that require broader changes (e.g. maybe better as a separate follow-up)?

Done.

> Regarding the outdated JS dependencies, we should fix this, but need to keep in mind that it doesn't always make sense to have the latest Gutenberg packages, as we need to be backward-compatible with 5.0. For that purpose, it's probably safest to lock all Gutenberg packages to those that were deployed with WordPress 5.0 for now.

Sure. I wasn't referring to Gutenberg packages though:

```
$ npm outdated
Package                                           Current   Wanted   Latest  Location
@babel/cli                                          7.2.3    7.5.0    7.5.0  googlesitekit
@babel/core                                         7.4.0    7.5.4    7.5.4  googlesitekit
@babel/plugin-transform-runtime                     7.4.0    7.5.0    7.5.0  googlesitekit
@babel/polyfill                                     7.4.0    7.4.4    7.4.4  googlesitekit
@babel/preset-env                                   7.4.2    7.5.4    7.5.4  googlesitekit
@material/button                                    2.1.1    2.3.0    3.0.0  googlesitekit
@material/checkbox                                  2.1.1    2.3.0    3.0.0  googlesitekit
@material/dialog                                    2.2.0    2.3.0    3.0.0  googlesitekit
@material/form-field                                2.1.1    2.3.0    3.0.0  googlesitekit
@material/layout-grid                              0.41.0   0.41.0    3.0.0  googlesitekit
@material/linear-progress                           1.1.0    1.1.0    3.0.0  googlesitekit
@material/list                                      2.1.1    2.3.0    3.0.0  googlesitekit
@material/menu                                      2.2.0    2.3.0    3.0.0  googlesitekit
@material/radio                                     2.1.1    2.3.0    3.0.0  googlesitekit
@material/react-select                             0.12.1   0.12.1   0.14.1  googlesitekit
@material/react-tab-bar                            0.12.0   0.12.0   0.14.1  googlesitekit
@material/react-text-field                         0.12.1   0.12.1   0.14.1  googlesitekit
@material/ripple                                    2.1.1    2.3.0    3.0.0  googlesitekit
@material/select                                    2.2.0    2.3.1    3.0.0  googlesitekit
@material/switch                                    2.1.1    2.3.0    3.0.0  googlesitekit
@material/textfield                                 2.1.1    2.3.1    3.0.0  googlesitekit
@material/theme                                     1.1.0    1.1.0    3.0.0  googlesitekit
@storybook/addon-viewport                          5.0.11    5.1.9    5.1.9  googlesitekit
@storybook/react                                   5.0.11    5.1.9    5.1.9  googlesitekit
@wordpress/api-fetch                                2.2.8    2.2.8    3.3.0  googlesitekit
@wordpress/browserslist-config                      2.3.0    2.5.0    2.5.0  googlesitekit
@wordpress/components                               7.2.0    7.4.0    8.0.0  googlesitekit
@wordpress/dom-ready                                2.2.0    2.4.0    2.4.0  googlesitekit
@wordpress/element                                  2.3.0    2.5.0    2.5.0  googlesitekit
@wordpress/hooks                                    2.2.0    2.4.0    2.4.0  googlesitekit
@wordpress/i18n                                     3.3.0    3.5.0    3.5.0  googlesitekit
@wordpress/library-export-default-webpack-plugin    1.1.0    1.3.0    1.3.0  googlesitekit
autoprefixer                                        9.5.0    9.6.1    9.6.1  googlesitekit
babel-eslint                                        9.0.0    9.0.0   10.0.2  googlesitekit
babel-loader                                        8.0.5    8.0.6    8.0.6  googlesitekit
backstopjs                                          3.8.8   3.9.13    4.1.9  googlesitekit
browser-sync                                       2.26.3   2.26.7   2.26.7  googlesitekit
bundlesize                                         0.17.1   0.17.2   0.18.0  googlesitekit
css-loader                                          1.0.1    1.0.1    3.0.0  googlesitekit
element-closest                                     2.0.2    2.0.2    3.0.1  googlesitekit
eslint                                             5.15.3   5.16.0    6.0.1  googlesitekit
eslint-config-standard                             11.0.0   11.0.0   13.0.0  googlesitekit
eslint-loader                                       2.1.2    2.2.1    2.2.1  googlesitekit
eslint-plugin-import                               2.16.0   2.18.0   2.18.0  googlesitekit
eslint-plugin-node                                  7.0.1    7.0.1    9.1.0  googlesitekit
eslint-plugin-promise                               3.8.0    3.8.0    4.2.1  googlesitekit
eslint-plugin-react                                7.12.4   7.14.2   7.14.2  googlesitekit
eslint-plugin-standard                              3.1.0    3.1.0    4.0.0  googlesitekit
focus-trap-react                                    5.0.1    5.0.1    6.0.0  googlesitekit
formdata-polyfill                                  3.0.17   3.0.18   3.0.18  googlesitekit
gulp                                                3.9.1    3.9.1    4.0.2  googlesitekit
gulp-filter                                         5.1.0    5.1.0    6.0.0  googlesitekit
gulp-imagemin                                       4.1.0    4.1.0    6.0.0  googlesitekit
gulp-livereload                                     3.8.1    3.8.1    4.0.1  googlesitekit
gulp-phpcs                                          2.2.0    2.2.0    3.0.0  googlesitekit
gulp-svgmin                                         2.1.0    2.2.0    2.2.0  googlesitekit
gulp-zip                                            4.2.0    4.2.0    5.0.0  googlesitekit
husky                                               1.3.1    1.3.1    3.0.0  googlesitekit
lint-staged                                         8.1.5    8.2.1    9.2.0  googlesitekit
lodash                                            4.17.11  4.17.14  4.17.14  googlesitekit
mini-css-extract-plugin                             0.4.5    0.4.5    0.7.0  googlesitekit
node-sass                                          4.11.0   4.12.0   4.12.0  googlesitekit
polyfill-library                                   3.31.1   3.35.0   3.35.0  googlesitekit
postcss-preset-env                                  6.6.0    6.7.0    6.7.0  googlesitekit
promise-polyfill                                    8.1.0    8.1.3    8.1.3  googlesitekit
puppeteer                                          1.13.0   1.18.1   1.18.1  googlesitekit
react                                              16.8.5   16.8.6   16.8.6  googlesitekit
react-dom                                          16.8.5   16.8.6   16.8.6  googlesitekit
start-server-and-test                              1.7.13    1.9.1    1.9.1  googlesitekit
stylelint                                          9.10.1   9.10.1   10.1.0  googlesitekit
stylelint-config-wordpress                         13.1.0   13.1.0   14.0.0  googlesitekit
stylelint-order                                     1.0.0    1.0.0    3.0.0  googlesitekit
stylelint-scss                                      3.5.4    3.9.1    3.9.1  googlesitekit
url-polyfill                                        1.1.5    1.1.6    1.1.6  googlesitekit
webpack                                            4.28.4   4.28.4   4.35.3  googlesitekit
webpack-cli                                         3.3.0    3.3.5    3.3.5  googlesitekit
webpackbar                                          3.1.5    3.2.0    3.2.0  googlesitekit
```

**Comment by felixarntz** [comment]
@adamsilverstein Would be great to have your review on this as well.

**Comment by adamsilverstein** [comment]
This is excellent, great work @swissspidy! 

When I ran the tests locally tests passed, however I did notice one error output, is this expected?

`  console.error tests/e2e/config/bootstrap.js:121
    Failed to load resource: the server responded with a status of 404 (Not Found)`

image

**Comment by swissspidy** [comment]
> When I ran the tests locally tests passed, however I did notice one error output, is this expected?

This error is just the error you see in the browser console when visiting the plugin's setup screen (i.e. http://localhost:9002/wp-admin/admin.php?page=googlesitekit-splash). It's not an error in the tests themselves.

According to Felix it's a known thing that the plugin's REST API endpoint returns a 404 on the setup screen, hence that error in the console.

**Comment by adamsilverstein** [comment]
@swissspidy @felixarntz Since this merge I'm seeing an issue locally when opening a module's settings, Site Kit->Settings-> then click Edit on a module. @lucymtc also confirmet this broken in her local.

image

Console errors:

image

**Comment by swissspidy** [comment]
@adamsilverstein Will look into this right now!

**Comment by lucymtc** [comment]
@swissspidy I tested with babel core ^7.5.4 and seems to fix it, can you test on your end?

**Comment by swissspidy** [comment]
@lucymtc Indeed that seems to help. I opened a PR at #280 to address this. It updates all the packages to ensure that they all work well together in their intended combination.

**Pull request merged**